loader-proto.c

```diff
@@ -0,0 +1,411 @@
+	   when asking for the buffer size with Buffer=NULL. I am assuming EFI_SUCCESS
+	   and EFI_BUFFER_TOO_SMALL are the only reasonable interpretations. */
+	if (EFI_ERROR(status) && status != EFI_BUFFER_TOO_SMALL) {
+		status = EFI_LOAD_ERROR;
+		goto out;
+	}
+
+	/* allocate buffer */
+	bprop->buffer = AllocatePool(bprop->size);
+	if (!bprop->buffer) {
+		status = EFI_OUT_OF_RESOURCES;
+		goto out;
+	}
+
+	/* read file */
+	status = lf2->LoadFile(lf2, bprop->dp, /*BootPolicy=*/false, &bprop->size, bprop->buffer);
+	if (EFI_ERROR(status))
+		goto out;
+
+out:
+	if (EFI_ERROR(status) && bprop->buffer)
+		FreePool(bprop->buffer);
+	return status;
+}
+
+static EFI_STATUS EFIAPI
+shim_load_image(BOOLEAN BootPolicy, EFI_HANDLE ParentImageHandle,
+                EFI_DEVICE_PATH *FilePath, VOID *SourceBuffer,
+                UINTN SourceSize, EFI_HANDLE *ImageHandle)
+{
+	SHIM_LOADED_IMAGE *image;
+	EFI_STATUS efi_status;
+	buffer_properties_t bprop;
+
+	if (BootPolicy)
+		return EFI_UNSUPPORTED;
+
+	if (!SourceBuffer || !SourceSize) {
+		if (try_load_from_sfs(FilePath, &bprop) == EFI_SUCCESS)
+			;
+		else if (try_load_from_lf2(FilePath, &bprop) == EFI_SUCCESS)
+			;
+		else
+			/* no buffer given and we cannot load from this device */
+			return EFI_LOAD_ERROR;
+
+		SourceBuffer = bprop.buffer;
+		SourceSize = bprop.size;
+	} else {
+		bprop.buffer = NULL;
+		/* even if we are using a buffer, try populating the
+		 * device_handle and file_path fields the best we can */
+		bprop.dp = FilePath;
+		efi_status = BS->LocateDevicePath(&EFI_DEVICE_PATH_GUID,
+		                                  &bprop.dp,
+		                                  &bprop.hnd);
+		if (efi_status != EFI_SUCCESS) {
+			/* can't seem to pull apart this DP */
+			bprop.dp = FilePath;
+			bprop.hnd = NULL;
+		}
+
+	}
+
+	image = AllocatePool(sizeof(*image));
+	if (!image) {
+		efi_status = EFI_OUT_OF_RESOURCES;
+		goto free_buffer;
+	}
+
+	SetMem(image, sizeof(*image), 0);
+
+	image->li.Revision = 0x1000;
+	image->li.ParentHandle = ParentImageHandle;
+	image->li.SystemTable = systab;
+	image->li.DeviceHandle = bprop.hnd;
+	image->li.FilePath = DuplicateDevicePath(bprop.dp);
+	image->loaded_image_device_path = DuplicateDevicePath(FilePath);
+
+	efi_status = handle_image(SourceBuffer, SourceSize, &image->li,
+	                          &image->entry_point, &image->alloc_address,
+	                          &image->alloc_pages);
+	if (EFI_ERROR(efi_status))
+		goto free_image;
+
+	*ImageHandle = NULL;
+	efi_status = BS->InstallMultipleProtocolInterfaces(ImageHandle,
+					&SHIM_LOADED_IMAGE_GUID, image,
+					&EFI_LOADED_IMAGE_GUID, &image->li,
+					&EFI_LOADED_IMAGE_DEVICE_PATH_GUID,
+					image->loaded_image_device_path,
+					NULL);
+	if (EFI_ERROR(efi_status))
+		goto free_alloc;
+
+	if (bprop.buffer)
+		FreePool(bprop.buffer);
+
+	return EFI_SUCCESS;
+
+free_alloc:
+	BS->FreePages(image->alloc_address, image->alloc_pages);
+free_image:
+	FreePool(image);
```

Comment:
We should also FreePool(image->li.FilePath) and FreePool(image->loaded_image_device_path).